src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py

```diff
@@ -1512,3 +1512,42 @@ def list_consumption_locations():
     client = web_client_factory()
     regions = client.list_geo_regions(sku='Dynamic')
     return [{'name': x.name.lower().replace(" ", "")} for x in regions]
+
+
+def enable_zip_deploy(resource_group_name, name, src, slot=None):
+    # if src is not of .zip extension throw exception
+    if not src.endswith('.zip'):
+        raise CLIError("src '{}' path needs to be .zip file".format(src))
+
+    client = web_client_factory()
+    user_name, password = _get_site_credential(client, resource_group_name, name)
+    scm_url = _get_scm_url(resource_group_name, name, slot)
+    zip_url = scm_url + '/api/zipdeploy'
+
+    import urllib3
+    try:
+        import urllib3.contrib.pyopenssl
+        urllib3.contrib.pyopenssl.inject_into_urllib3()
+    except ImportError:
+        pass
+
+    headers = urllib3.util.make_headers(basic_auth='{0}:{1}'.format(user_name, password))
+    headers['content-type'] = 'application/octet-stream'
+
+    import requests
+    # Read file content
+    fs = open(src, 'rb')
+    zip_content = fs.read()
+
+    try:
+        r = requests.post(zip_url, data=zip_content, headers=headers)
```

Comment:
`requests.post` support sending a file directly, why use that feature? I'm not expert regarding `requests` but I would assume the approach you adopt here will load the entire file into the memory before sends it out. Given that you do not modify the content, it is wasteful.

Reference:
* http://docs.python-requests.org/en/master/user/advanced/#streaming-uploads
* http://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file

Reply:
I am confused here on the ask here - the API itself takes the content of a zip file  - it doesn't support sending chunks of data. Are you suggesting that we extract all files from the zip & use the requests.post to send  data using files =extractedFiles?

---
In reply to: [145754064](https://github.com/Azure/azure-cli/pull/4700#discussion_r145754064) [](ancestors = 145754064)

Reply:
Please refer to the content in the first link. You're read entire zip file content into the memory and send it through HTTP. The `requests` package allows you set the file stream as to the data parameter of the `post` function. It will save memory and keep your code concise. Using context manager also helps you close the file automatically.

```
try:
    with open(src, 'rb') as f:
        r = requests.post(zip_url, data=f, headers=headers)
except applicable_exceptions:
    ....
```